### PR TITLE
Fix commit hook for older clang-format versions

### LINKED ISSUE #5249: git-clang-format `--diffstat` option added in clang-14
Versions of clang before 14 don't have the `--diffstat` option to `git-clang-format` (https://reviews.llvm.org/D84375, version tags listed [here](https://github.com/llvm/llvm-project/commit/191a395343b984d26d058c449cefe1a9ac0ff927#diff-b36a1ef270f7220995ade6ac916bed21677f2d5e3c5d1131052bba09e2ab92d3L506)), yet we support all the way back to clang-10. Clang-10 is what ships with ubuntu 20.04 and so by default our git hooks are broken on that (this is what I use).

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Fixes #5249 . @knelli2 could you confirm?

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: tools/Hooks/pre-commit.sh [hunk 1]
```diff
@@ -36,17 +36,14 @@ printf '%s\0' "${commit_files[@]}" | \
 # Use git-clang-format to check for any suspicious formatting of code.
 @CLANG_FORMAT_BIN@ --version > /dev/null
 if [ $? -eq 0 ]; then
-    clang_format_diffstat=$(@GIT_EXECUTABLE@ --no-pager \
-        clang-format --binary @CLANG_FORMAT_BIN@ --diffstat --quiet)
+    clang_format_diff=$(@GIT_EXECUTABLE@ --no-pager \
+        clang-format --binary @CLANG_FORMAT_BIN@ --diff --quiet)
```
Comment (knelli2) [inline tools/Hooks/pre-commit.sh:40]: Even with `--quiet`, if there are no changes this returns

```
no modified files to format
```

File: tools/Hooks/pre-commit.sh [hunk 2]
```diff
@@ -36,17 +36,14 @@ printf '%s\0' "${commit_files[@]}" | \
...
+    # so we check the diff output instead (see issue:
     # https://github.com/llvm/llvm-project/issues/54758)
-    if [ -n "$clang_format_diffstat" ]; then
-        found_error=1
-        @GIT_EXECUTABLE@ clang-format --binary @CLANG_FORMAT_BIN@ --diff \
-            > @CMAKE_SOURCE_DIR@/.clang_format_diff.patch
-        echo "Found C++ formatting errors:"
-        echo "$clang_format_diffstat"
+    if [ -n "$clang_format_diff" ]; then
+        echo $clang_format_diff > @CMAKE_SOURCE_DIR@/.clang_format_diff.patch
```
Comment (knelli2) [inline tools/Hooks/pre-commit.sh]: Need quotes around `$clang_format_diff` to preserve newlines and have the patch file formatted properly

**Review by knelli2 — CHANGES_REQUESTED** [review]

File: tools/Hooks/pre-commit.sh (hunk 1 above)
Comment (nilsvu) [reply tools/Hooks/pre-commit.sh:40]: Ugh this works with clang-format v15

Comment (knelli2) [reply tools/Hooks/pre-commit.sh:40]: Annoying that git-clang-format changed their behavior of `--quiet`

**Comment by nilsvu** [comment]
Does this work with clang-10 now @knelli2?

File: tools/Hooks/pre-commit.sh [hunk 3]
```diff
@@ -36,17 +36,16 @@ printf '%s\0' "${commit_files[@]}" | \
...
-        echo "Found C++ formatting errors:"
-        echo "$clang_format_diffstat"
+    if [ -n "$clang_format_diff" ] \
+        && [[ "$clang_format_diff" != *"no modified files to format"* ]] \
+        && [[ "$clang_format_diff" != *"did not modify any files"* ]]; then
+        echo "$clang_format_diff" > @CMAKE_SOURCE_DIR@/.clang_format_diff.patch
+        echo "Found C++ formatting errors."
         echo "Please run 'git clang-format' in the repository."
         echo "You can also apply the patch directly:"
         echo "git apply @CMAKE_SOURCE_DIR@/.clang_format_diff.patch"
```
Comment (knelli2) [inline tools/Hooks/pre-commit.sh:51]: Can you add an `echo` after this last line so it separates the formatting error message from the actual commit message? Makes it easier to read.

**Review by knelli2 — CHANGES_REQUESTED** [review]
Yes now this works, just one small thing.

File: tools/Hooks/pre-commit.sh (hunk 3 above)
Comment (nilsvu) [reply tools/Hooks/pre-commit.sh:51]: ok done

**Review by knelli2 — APPROVED** [review]

**Pull request merged**